implement determining the threshold pressure from the initial condition

andlaus opened the pull request:
This needs to be done if a equilibration region transition is
mentioned by the THPRES keyword, but no value is given for this record
in the third item. (it seems that this is used quite frequently.)

Also, the approach taken by this patch also does not collide with the
restart machinery as far as I can see. This is because the initial
condition is applied by the simulator before the state at the restart
time is loaded. (I interpreted the code that way, but I could be
wrong, could anyone verify this?)

since it is pretty elaborate to calculate initial condition, this
patch is pretty messy. I also do not know if Eclipse does include
capillary pressure in this calculation or not (this patch does) and
there may be other bugs lurking. happy testing!

atgeirr commented:
Thank you for the patch(es)! I intend to review and merge them eventually, but that will wait until after the release, unless there are significant reasons to make an exception for this (new variant) feature.

andlaus commented:
yeah, I agree. I forgot to mention that I consider this to be post-release stuff.

andlaus commented:
I suppose the same argument as for OPM/opm-material#81 applies here: the release is not affected anymore, so the PR can savely be merged (after review of course).

atgeirr commented:
> the release is not affected anymore, so the PR can savely be merged (after review of course).

Intend to review after release.

totto82 commented:
I have compared the results with Eclipse. The results does not match. I will send you an email with the comparison.

totto82 commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,7 +68,154 @@ namespace Opm
+            for (int phaseIdx = 0; phaseIdx < numPhases; ++phaseIdx) {
+                phasePressure[phaseIdx].resize(numCells);
+                for (int cellIdx = 0; cellIdx < numCells; ++cellIdx) {
+                    phasePressure[phaseIdx][cellIdx] = initialState.pressure()[cellIdx] + (capPress[cellIdx*numPhases + phaseIdx] - capPress[cellIdx*numPhases + refPhaseIdx]);
```
Is the sign correct for the water phase?

andlaus replied:
I'll have to check. (will do this tomorrow.) for opm-material it would be correct. opm-core is a different beast, though...

totto82 commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,7 +68,154 @@ namespace Opm
+                    rho[opos][cellIdx] = surfaceDensity[pvtRegion[cellIdx]][opos]*b[cellIdx];
+
+                    if (pu.phase_used[BlackoilPhases::Vapour]) {
+                        rho[opos][cellIdx] +=
```
Shouldn't it be: 
```rho[opos][cellIdx] += surfaceDensity[pvtRegion[cellIdx]][gpos] *
initialState.gasoilratio()[cellIdx]*b[cellIdx];```

andlaus replied:
true.

totto82 commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,7 +68,154 @@ namespace Opm
+
+                    if (pu.phase_used[BlackoilPhases::Vapour]) {
+                        rho[gpos][cellIdx] +=
+                            initialState.rv()[cellIdx]*b[cellIdx];
```
surfaceDensity[pvtRegion[cellIdx]][opos] is missing ?

totto82 commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -74,12 +232,57 @@ namespace Opm
+                const int gc2 = (gc == 0) ? c2 : gc[c2];
+                const int eq1 = eqlnumData[gc1];
+                const int eq2 = eqlnumData[gc2];
+
```
I think only phases present in the upstream cell should be considered?

andlaus replied:
for gravity correction you also need to calculate the density of the downstream cell, right?

totto82 replied:
What I mean is something like: 
  ```
double s1 = initialState.saturation()[numPhases*c1 + phaseIdx];
double s2 = initialState.saturation()[numPhases*c2 + phaseIdx];
if ( (p1 > p2 && s1 > 0) || (p2 > p1 && s2 > 0) ) {
    maxDp[barrierId] = std::max(maxDp[barrierId], std::abs(p1 - p2));
 }
```

andlaus replied:
Ok, I see what you mean. if the density contains a meaningful value it should work as is I think.

Also, only doing this if there is saturation in the upstream cell means that the phase is discarded when it comes to calculating threshold pressure. is that desireable? (To me be both ways seem to be meaningful physically, but what does E100?)

totto82 replied:
Some more reverse engineering point in the direction of using smin instead of 0

totto82 replied:
According to the manual differences in the oil phase should not effect the thpress in the water zone and so on. I guess the only way to figure this out is to test it.

andlaus replied:
ok, I've updated the PR to consider the saturation of the upstream cell and use the average phase density.

totto82 replied:
Based on testing, I think this is more correct. 
```
 double smin1 = smin[numPhases*c1 + phaseIdx];
 double smin2 = smin[numPhases*c2 + phaseIdx];

 if ( (p1 > p2 && s1 > smin1) || (p2 > p1 && s2 > smin2) ) { ... } 
```

andlaus replied:
ok, changed.

andlaus commented:
ok, I added a fixup for the missing surface densities and the capillary sign of the capillary pressure.

totto82 commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -74,12 +245,57 @@ namespace Opm
+                    double p1 = phasePressure[phaseIdx][c1];
+                    double p2 = phasePressure[phaseIdx][c2];
+                    p1 += rho[phaseIdx][c1]*gravity*(z1 - zAvg);
+                    p2 += rho[phaseIdx][c2]*gravity*(z2 - zAvg);
```
The average density should be used for both sides (as in the simulator).
Is the sign wrong?

andlaus replied:
hm, I'm not sure if that's a good idea: all values (this includes density) are constant within FV cells, right? is there an effect if this is changed for the conservation equations?

andlaus replied:
also, the sign should be correct: if the depth of the cell is larger than the average, the pressure is increased, if it is smaller, pressure gets decreased. (for positive gravity constants, which seems to be the case in flow.)

totto82 replied:
The purpose of the thpress is to avoid flow between regions if not an external force is present. Therefore I will argue that we should use the same formulation as computeMassFlux in BlackoilModeBase_impl.hpp.

andlaus replied:
ok, that's an argument in favour of the average density. (agreed.) But the sign should be correct?

totto82 commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,7 +68,167 @@ namespace Opm
+                    if (pu.phase_used[BlackoilPhases::Vapour]) {
+                        int gpos = pu.phase_pos[BlackoilPhases::Vapour];
+                        rho[opos][cellIdx] +=
+                            surfaceDensity[pvtRegion[cellIdx]][gpos]*initialState.surfacevol()[cellIdx]*b[cellIdx];
```
This is still wrong.

andlaus replied:
*arg*: the names `surfacevolume()` and `gasoilratio()` are  actually worse than `foo()` and `bar()` would be: for the latter you'd actually be forced to look up what they mean instead of being mislead by the method name.

anyway, fixed. thanks.

andlaus commented:
ok, I've squashed the fixups and rebased it on top of the latest master version. If we're lucky, it can be merged now, but @totto82 will do some more tests.

totto82 commented:
I am able to reproduce Eclipse with this code, but then I have to use       
UgGridHelpers::cellCenterDepth(G, cell) instead of            
UgGridHelpers::cellCentroidCoordinate(G, cell, 2) 
in the initialization as well. I think cellCenterDepth is more similar to Eclipse and should be used in the initialization. I will test it and make a separate PR for this.

atgeirr commented:
Since the release is done, I have now looked at this.

I cannot argue with the fact that this code is tested to work correctly, if I understand @totto82. It does look like it contains a lot of pasted code from various places, though. I initially believed that the output from the initialization process would be enough (that the  fluid object for example would not be needed), I assume you have verified that such an approach would not be sufficient.

Given that it *is* needed, we must approach this as it is done in the model, otherwise we risk inconsistencies. But duplicating all that is needed for the calculations is a big and error-prone thing. Would it then not be far better to move this part to the model class itself, where the same function could be used for both purposes (determining the threshold pressures as well as the phase head differences and resulting fluxes)? This could also make the handling of NNCs easier.

andlaus commented:
> I initially believed that the output from the initialization process would be enough (that the fluid object for example would not be needed), I assume you have verified that such an approach would not be sufficient.

I have indeed verified that it is insufficient. (because I was quite annoyed by the fact that the phase densities need to be re-calculated from scratch at any place where they are required ;).) Long-term, I think it would be beneficial to split the "reservoir state" into a "primal state" which features only the primary variables on which the non-linear and linear solvers operate on and a "secondary state" which provides all quantities required for calculating the residual that are derived from the "primal" quantities. The "secondary state" would indeed be sufficient for this functionality.
 
> Given that it is needed, we must approach this as it is done in the model, otherwise we risk inconsistencies. Would it then not be far better to move this part to the model class itself, where the same function could be used for both purposes (determining the threshold pressures as well as the phase head differences and resulting fluxes)? This could also make the handling of NNCs easier.

I certainly agree that this patch is messy and prone to inconsistencies, but in practice your suggestion implies moving the `thresholdPressures()` functions out of opm-core into opm-autodiff and -- probably more relevantly -- as far as I can see the duplication would not vanish but just move to a different module. (because gravity correction is "hard coded" into the assembly process and thus not directly available for the initialization code.)

Finally, the code duplication and consistency arguments equally apply to the `initStateEquil()` functions in opm-core?

atgeirr commented:
Thanks for your explanations, @andlaus. The discussion of reservoir state is interesting, I'll not continue that here however. Your comment about initStateEquil() also being a consistency risk makes me think that perhaps the right thing to do in the long term is to move all these things into the model. With different models, the equations are different and everything that depends on the equations (including initialization) should therefore be part of the model (system, if not class).

I'll review the code, with intention to merge.

andlaus commented:
> Your comment about initStateEquil() also being a consistency risk makes me think that perhaps the right thing to do in the long term is to move all these things into the model.

basically that's another argument of abandoning opm-core ;)

> With different models, the equations are different and everything that depends on the equations (including initialization) should therefore be part of the model (system, if not class).

that's the reason why opm-material has the "fluid state" concept: using this, each model can make whatever assumptions are the most appropriate for it, but generic code does not need to care about these details. maybe you should have a closer look to see if something similar would be beneficial for opm-autodiff.

atgeirr commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,10 +68,186 @@ namespace Opm
             auto eqlnumData = eqlnum->getData();
 
-            // Set values for each face.
+            int numPhases = initialState.numPhases();
```
I prefer that such variables are const, I've gotten so used to it that when they are *not* const I look for where they are changed!

atgeirr commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,10 +68,186 @@ namespace Opm
+            std::vector<int> pvtRegion(numCells);
+            const auto& cartPvtRegion = eclipseState->getIntGridProperty("PVTNUM")->getData();
+            for (int cellIdx = 0; cellIdx < numCells; ++cellIdx) {
+                int cartCellIdx = gc?gc[cellIdx]:cellIdx;
```
Please make the variable const, and add spaces around `?` and `:`.

atgeirr commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,10 +68,186 @@ namespace Opm
+            // the inverse formation volume factors
+            std::vector<PhasePresence> cond(numCells);
+            for (int cellIdx = 0; cellIdx < numCells; ++cellIdx) {
+                if (pu.phase_used[BlackoilPhases::Aqua] && initialState.saturation()[numPhases*cellIdx + pu.phase_pos[BlackoilPhases::Aqua]] > 0.0) {
```
These lines are unreadably long, so I misunderstood a little at first. A reference alias or twou should make it much easier to read.

atgeirr commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,10 +68,186 @@ namespace Opm
+
+                if (pu.phase_used[BlackoilPhases::Aqua]) {
+                    int wpos = pu.phase_pos[BlackoilPhases::Aqua];
+                    phasePressure[wpos][cellIdx] = initialState.pressure()[cellIdx] + (capPress[cellIdx*numPhases + opos] - capPress[cellIdx*numPhases + wpos]);
```
Again, very long lines!

atgeirr commented on opm/core/utility/thresholdPressures.hpp:
```diff
@@ -57,10 +68,186 @@ namespace Opm
+
+            for (int cellIdx = 0; cellIdx < numCells; ++ cellIdx) {
+                assert(pu.phase_used[BlackoilPhases::Liquid]); // we currently hard-code the oil phase as the reference phase!
+                int opos = pu.phase_pos[BlackoilPhases::Liquid];
```
Could (and therefore should) be const.

atgeirr commented:
While I have made a few nitpicky remarks as usual, I find the code good in general. There is one change I will request though: to put the computation of maxDp into its own function. That will allow the (simple) logic of the thresholdPressure() function to be obvious to the reader (now it is at the bottom of a long function).

andlaus commented:
ok, all variables have been `const`ified in that functions and long lines have been split.

atgeirr commented:
Still missing separation of maxDp calculation into its own function.

andlaus commented:
> Still missing separation of maxDp calculation into its own function.

this took a few minutes longer. now it's done. don't forget to also merge OPM/opm-autodiff#506 and OPM/opm-polymer#151

atgeirr commented:
Thanks for the work, merging.
